File: js_backup/formatter.js
export function formatToolResult(query, toolResult) {
    // Fast template-based formatting — no LLM call needed
    if (toolResult.status === "error") {
        return `❌ Error: ${toolResult.message}`;
    }

    const { data, message } = toolResult;

    // Simple scalar results (time, date, calculator)
    if (data === null || data === undefined) {
        return message || "Done.";
    }
    if (typeof data === "string" || typeof data === "number") {
        return `${message}\n\n${data}`;
    }

    // Array of objects (database rows, search results)
    if (Array.isArray(data)) {
        if (data.length === 0) return `${message || "No results found."}`;

        // Format as a readable list
        const keys = Object.keys(data[0]);
        const rows = data.map((row, i) =>
            `${i + 1}. ${keys.map(k => `${k}: ${row[k]}`).join(" | ")}`
        ).join("\n");
        return `${message}\n\n${rows}`;
    }

    // Single object
    if (typeof data === "object") {
        const entries = Object.entries(data)
            .map(([k, v]) => `• ${k}: ${v}`)
            .join("\n");
        return `${message}\n\n${entries}`;
    }

    return message || JSON.stringify(data);
}

File: server/formatter.py
import json
from typing import Dict, Any, Union, List

def format_tool_result(query: str, tool_result: Dict[str, Any]) -> str:
    if tool_result.get("status") == "error":
        return f"❌ Error: {tool_result.get('message', 'Unknown error')}"
        
    data = tool_result.get("data")
    message = tool_result.get("message", "Done.")
    
    if data is None:
        return message
        
    if isinstance(data, (str, int, float, bool)):
        return f"{message}\n\n{data}"
        
    if isinstance(data, list):
        if len(data) == 0:
            return message if tool_result.get("message") else "No results found."
            
        if isinstance(data[0], dict):
            keys = data[0].keys()
            rows = []
            for i, row in enumerate(data):
                row_str = " | ".join([f"{k}: {row.get(k, '')}" for k in keys])
                rows.append(f"{i + 1}. {row_str}")
            return f"{message}\n\n" + "\n".join(rows)
            
        return f"{message}\n\n" + "\n".join([f"• {item}" for item in data])
        
    if isinstance(data, dict):
        entries = "\n".join([f"• {k}: {v}" for k, v in data.items()])
        return f"{message}\n\n{entries}"
        
    return message or json.dumps(data)
